microovn/src/charm.py

```diff
@@ -0,0 +1,153 @@
+        framework.observe(self.on.leader_elected , self._on_leader_elected)
+        framework.observe(self.on[WORKER_RELATION].relation_changed,
+                            self._on_cluster_changed)
+        framework.observe(self.on[WORKER_RELATION].relation_created,
+                            self._handle_relation_created)
+
+
+    def add_hostname(self, relation_name):
+        relation = self.model.get_relation(relation_name)
+        relation.data[self.unit]["hostname"] = get_hostname()
+
+    def find_token(self, relation_name):
+        if not (relation := self.model.get_relation(relation_name)):
+            return False
+
+        tokens = {
+            token
+            for unit in relation.units
+            if (token := relation.data[unit].get(mirrorid(get_hostname()))) and \
+            token != "empty"
+        }
+        if len(tokens) > 1:
+            self.unit.status = ops.MaintenanceStatus("Too many tokens")
+            return False
+        if len(tokens) == 0:
+            self.unit.status = ops.MaintenanceStatus("No token yet")
+            return False
+
+        logger.info("found token")
+        (token, ) = tokens
+        return token
+
+    def join_with_token(self, token):
+        self.unit.status = ops.MaintenanceStatus("Joining cluster")
+        os.system("microovn cluster join {}".format(token))
+        self._stored.in_cluster=True
+        self.unit.status = ops.ActiveStatus("Joined cluster")
+
+    def _on_install(self, event: ops.InstallEvent):
+        self.unit.status = ops.MaintenanceStatus("installing microovn snap")
+        os.system("snap install microovn --channel latest/edge")
+        self.unit.status = ops.MaintenanceStatus("waiting for microovn ready")
+        os.system("microovn waitready")
+        self.unit.status = ops.MaintenanceStatus("waiting for cluster setup")
+
+    def _on_start(self, event: ops.StartEvent):
+        if self.unit.is_leader():
+            # only bootstrap if leader
+            # TODO: fix cluster bootstrapped, leader changes to new node not in
+            #       cluster, bootstrapps new cluster can be fixed via having some
+            #       other characteristic determine who manages this side of the
+            #       mirror ie: microcluster leader
+            os.system(f"microovn cluster bootstrap")
+            self._stored.in_cluster = True
+            self.unit.status = ops.ActiveStatus("cluster bootstrapped")
+
+    def _handle_relation_created(self, event: ops.RelationCreatedEvent):
+        self.add_hostname(WORKER_RELATION)
+        if self.unit.is_leader():
+            self.update_tokens(WORKER_RELATION)
+
+    def _on_remove(self, event: ops.RemoveEvent):
+        if self._stored.in_cluster:
+            os.system("microovn cluster remove {}".format(get_hostname()))
+
+    def _handle_relation_joined(self, relation_name):
+        if not self.unit.is_leader() or not self._stored.in_cluster:
```

Comment:
This method's name suggests that it's tied to and observed "join" event ( like other`_handle_` methods). However it is not. It's called only from `_on_cluster_change`, so it doesn't offer any code reuse. It's contents can be merged to `_on_cluster_changed` to improve code readability.